netbsd/Platform.c

```diff
@@ -0,0 +1,327 @@
+   &BatteryMeter_class,
+   &HostnameMeter_class,
+   &SysArchMeter_class,
+   &AllCPUsMeter_class,
+   &AllCPUs2Meter_class,
+   &AllCPUs4Meter_class,
+   &AllCPUs8Meter_class,
+   &LeftCPUsMeter_class,
+   &RightCPUsMeter_class,
+   &LeftCPUs2Meter_class,
+   &RightCPUs2Meter_class,
+   &LeftCPUs4Meter_class,
+   &RightCPUs4Meter_class,
+   &LeftCPUs8Meter_class,
+   &RightCPUs8Meter_class,
+   &BlankMeter_class,
+   NULL
+};
+
+void Platform_init(void) {
+   /* no platform-specific setup needed */
+}
+
+void Platform_done(void) {
+   /* no platform-specific cleanup needed */
+}
+
+void Platform_setBindings(Htop_Action* keys) {
+   /* no platform-specific key bindings */
+   (void) keys;
+}
+
+int Platform_getUptime() {
+   struct timeval bootTime, currTime;
+   const int mib[2] = { CTL_KERN, KERN_BOOTTIME };
+   size_t size = sizeof(bootTime);
+
+   int err = sysctl(mib, 2, &bootTime, &size, NULL, 0);
+   if (err) {
+      return -1;
+   }
+   gettimeofday(&currTime, NULL);
+
+   return (int) difftime(currTime.tv_sec, bootTime.tv_sec);
+}
+
+void Platform_getLoadAverage(double* one, double* five, double* fifteen) {
+   struct loadavg loadAverage;
+   const int mib[2] = { CTL_VM, VM_LOADAVG };
+   size_t size = sizeof(loadAverage);
+
+   int err = sysctl(mib, 2, &loadAverage, &size, NULL, 0);
+   if (err) {
+      *one = 0;
+      *five = 0;
+      *fifteen = 0;
+      return;
+   }
+   *one     = (double) loadAverage.ldavg[0] / loadAverage.fscale;
+   *five    = (double) loadAverage.ldavg[1] / loadAverage.fscale;
+   *fifteen = (double) loadAverage.ldavg[2] / loadAverage.fscale;
+}
+
+int Platform_getMaxPid() {
+   // https://nxr.netbsd.org/xref/src/sys/sys/ansi.h#__pid_t
+   // pid is assigned as a 32bit Integer.
+   return INT32_MAX;
+}
+
+double Platform_setCPUValues(Meter* this, int cpu) {
+   const NetBSDProcessList* npl = (const NetBSDProcessList*) this->pl;
+   const CPUData* cpuData = &npl->cpus[cpu];
+   double total = cpuData->totalPeriod == 0 ? 1 : cpuData->totalPeriod;
+   double totalPercent;
+   double* v = this->values;
+
+   v[CPU_METER_NICE] = cpuData->nicePeriod / total * 100.0;
+   v[CPU_METER_NORMAL] = cpuData->userPeriod / total * 100.0;
+   if (this->pl->settings->detailedCPUTime) {
+      v[CPU_METER_KERNEL]  = cpuData->sysPeriod / total * 100.0;
+      v[CPU_METER_IRQ]     = cpuData->intrPeriod / total * 100.0;
+      v[CPU_METER_SOFTIRQ] = 0.0;
+      v[CPU_METER_STEAL]   = 0.0;
+      v[CPU_METER_GUEST]   = 0.0;
+      v[CPU_METER_IOWAIT]  = 0.0;
+      v[CPU_METER_FREQUENCY] = NAN;
+      this->curItems = 8;
+      totalPercent = v[0] + v[1] + v[2] + v[3];
+   } else {
+      v[2] = cpuData->sysAllPeriod / total * 100.0;
+      v[3] = 0.0; // No steal nor guest on NetBSD
+      totalPercent = v[0] + v[1] + v[2];
+      this->curItems = 4;
+   }
+
+   totalPercent = CLAMP(totalPercent, 0.0, 100.0);
+
+   v[CPU_METER_TEMPERATURE] = NAN;
+
+   return totalPercent;
+}
+
+void Platform_setMemoryValues(Meter* this) {
```

Comment:
Looking at the help screen this normally should be 4 values: `used`/`buffers`/`shared`/`cached`. Could you please re-check the memory meter really just want's those three values / could you add the `shared` value? Note: `shared` includes stuff like tmpfs on Linux for usability reasons, that are normally accounted as `cached`.

Reply:
This was one of the areas I spent quite sometime trying to figure out how things work, the current implementation is based on the values exposed by [uvm(9)'s sysctl interface](https://man.netbsd.org/uvm.9#SYSCTL), uvm is the virtual memory system of NetBSD. I am not sure how `shared` would be defined in this situation, I shall ask around a bit in the NetBSD developers circle and get back to you on this.

Reply:
TIA. Otherwise assume `0` for now.

Reply:
I looked up the other BSD implementation and followed the same over here now. They do not set the `shared` value. 

Should be done in a3140cd